### PR TITLE
Feature/variant helpers

### PR DESCRIPTION
**Pull request by rfhaque**
Merge #356 first.  Closes #400.

Programming model and modifier.

### CODE REVIEW

File: lib/benchpark/experiment.py [hunk 1]
```diff
@@ -65,30 +114,124 @@ def compute_config_section(self):
...
+
+    def add_spack_spec(self, package_name, spec=None):
+        if spec:
+            self.package_specs[package_name] = {
+                "pkg_spec": spec[0],
+                "compiler": spec[1],
+            }
+        else:
+            self.package_specs[package_name] = {}
+
```
Comment (rfhaque) [inline lib/benchpark/experiment.py]: @scheibelp @pearce8 Refactored lib/experiment.py to remove much of the boilerplate code from the application experiment.py

File: lib/benchpark/experiment.py [hunk 2]
```diff
@@ -65,30 +114,124 @@ def compute_config_section(self):
...
+             )
+
+        spack_variants = [cls.get_spack_variants() for cls in self.helpers]
+        self.package_specs[self.name]["pkg_spec"] += " ".join(spack_variants+list(self.spec.variants["extra_spack_specs"])).strip()
+
+        return {
+            "packages": {k: v for k, v in self.package_specs.items() if v},
+            "environments": {self.name: {"packages": list(self.package_specs.keys())}},
+        }
+
```
Comment (rfhaque) [inline lib/benchpark/experiment.py]: @scheibelp @pearce8 Refactored lib/experiment.py to remove much of the boilerplate code from the application experiment.py

File: lib/benchpark/experiment.py [hunk 3]
```diff
@@ -65,30 +114,124 @@ def compute_config_section(self):
...
         return {
             "ramble": {
                 "include": self.compute_include_section(),
                 "config": self.compute_config_section(),
-                "modifiers": self.compute_modifiers_section(),
-                "applications": self.compute_applications_section(),
-                "software": self.compute_spack_section(),
+                "modifiers": self.compute_modifiers_section_wrapper(),
+                "applications": self.compute_applications_section_wrapper(),
+                "software": self.compute_spack_section_wrapper(),
```
Comment (rfhaque) [inline lib/benchpark/experiment.py:253]: @scheibelp @pearce8 Refactored lib/experiment.py to remove much of the boilerplate code from the application experiment.py

File: var/exp_repo/experiments/amg2023/experiment.py [hunk 4]
```diff
@@ -23,173 +21,111 @@ class Amg2023(Experiment):
...
             )
 
         # TODO: Support variant
         p = "2"
-        variables["px"] = p
-        variables["py"] = p
-        variables["pz"] = p
+        self.add_experiment_variable("px", p, True)
+        self.add_experiment_variable("py", p, True)
+        self.add_experiment_variable("pz", p, True)
```
Comment (rfhaque) [inline var/exp_repo/experiments/amg2023/experiment.py]: @scheibelp @pearce8 The application can use an API to set experiment variables and other constructs

**Review by pearce8 — APPROVED** [review]

**Pull request merged**